src/core/lib/event_engine/grpc_ares_wrapper.h

```diff
@@ -0,0 +1,332 @@
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_trace_ares_wrapper)) {                   \
+      gpr_log(GPR_INFO, "(EventEngine c-ares wrapper) " format, __VA_ARGS__); \
+    }                                                                         \
+  } while (0)
+
+// Base class of the c-ares based DNS lookup functionality. The actual concrete
+// leaf class represents each type of request i.e. A/AAAA, SRV, TXT.
+// GrpcAresRequest handles the logics to initialize and destroy the c-ares
+// channel (one channel for one request). Sets the name servers configuration
+// for the channel. And interacts with c-ares sockets/fds and gRPC poller. It
+// also handles logics to start and cancel timers.
+class GrpcAresRequest
+    : public grpc_core::InternallyRefCounted<GrpcAresRequest> {
+ public:
+  ~GrpcAresRequest() override;
+
+  // Cancels an inflight request, returns true if cancel succeeds and will start
+  // the shutdown process.
+  bool Cancel() ABSL_LOCKS_EXCLUDED(mu_);
+  void Orphan() override {}
+
+ protected:
+  GrpcAresRequest(absl::string_view name, EventEngine::Duration timeout,
+                  std::unique_ptr<GrpcPolledFdFactory> polled_fd_factory,
+                  EventEngine* event_engine);
+
+  // The workhorse function. Gets the live sockets/fds used by c-ares, creates
+  // FdNode if it is not currently tracked in fd_node_list_. Registers the
+  // socket with the poller for read and/or write based on c-ares's demand. And
+  // shutdown and destroys the poller handles whose sockets are no longer in use
+  // by c-ares.
+  // This function is called in every opportunities when there might be a change
+  // to c-ares' sockets for the channel.
+  void Work() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+  // Common logic to start the query timeout timer and the ares backup poll
+  // timer. This is only called in each leaf class' Start method.
+  void StartTimers() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+  // Common logic to cancel the query timeout timer and the ares backup poll
+  // timer. This is called when the request is cancelled or shutting down.
+  void CancelTimers() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+  // These 2 methods are deliberately thread-unsafe and should only be called in
+  // the factory methods of the leaf classes as part of initialization.
+  absl::StatusOr<std::string> ParseNameToResolve();
+  absl::Status InitializeAresOptions(absl::string_view dns_server);
+
+  grpc_core::Mutex mu_;
+  bool initialized_ ABSL_GUARDED_BY(mu_) = false;
+  // name to resolve
+  const std::string name_ ABSL_GUARDED_BY(mu_);
+  // ares channel
+  ares_channel channel_ ABSL_GUARDED_BY(mu_) = nullptr;
+  // host to resolve, parsed from name_
+  std::string host_ ABSL_GUARDED_BY(mu_);
+  bool cancelled_ ABSL_GUARDED_BY(mu_) = false;
+  bool shutting_down_ ABSL_GUARDED_BY(mu_) = false;
+  EventEngine* event_engine_;
+
+ private:
+  // A FdNode saves (not owns) a live socket/fd which c-ares creates and owns a
```

Comment:
small nit: Suggest adding a comma "... which c-ares creates*,* and owns ...". I had to read it twice to realize this did not mean c-ares creates and owns a GrpcPolledFd, which made no sense to me :-)

Reply:
Good point. Done.